### DISCUSSION TITLE
Deforming streamlines and transforming them to an aligned image space

### DISCUSSION THREAD

**Discussion by jcgriff1** [dipy/dipy#2646] [Registration]
Hello,

I have streamlines that I have computed in diffusion space, and I want to transform them into the space of the t1-weighted scan for visualization. I have computed transforms that (a) rigidly align the anatomical to DWI space, and (b) non-linearly warp the aligned anatomical to the DWI b0. 

I see that there is a `deform_streamlines()` function, but I'm finding the tutorial (https://dipy.org/documentation/1.5.0/examples_built/streamline_registration/#example-streamline-registration) very difficult to follow, and it's unclear how I should use this function for my intended purpose. 

Could someone clarify what the input arguments to deform_streamlines() should be in this case? I'm very unfamiliar with the DiPy nomenclature for registrations/spaces, so any help would be appreciated:

```
streamlines_wrp = deform_streamlines(streamlines = streamlines, # DWI-space tractography results loaded with TrkFile.load()
                       deform_field = diff_affine.get_backward_field(), # inverse of the deformation applied to the rigidly aligned T1
                       stream_to_current_grid = np.linalg.inv(trk_file_affine), # streamline to diffusion space?
                       current_grid_to_world = ? # What should this be?
                       stream_to_ref_grid = ? # What should this be?
                       ref_grid_to_world = ? # What should this be?
                       )
```

**Comment by skoudoro** [comment]
Hi @jcgriff1, 

Sorry for the late answer. I agree that this tutorial is confusing and we will try to update it for the next release.

I recommend reading this document: https://nipy.org/nibabel/coordinate_systems.html

Concerning your question for deform_streamlines

- streamlines=streamlines      # Streamlines are usually in the 3D world space 
- deform_field   # to obtain this deformation field, you needed 2 volumes/images, the reference image and the moving images. you need to get their affine matrix for both of them. those matrices will be use as the parameter below
- stream_to_current_grid    # scanner to grid coordinate of your moving image. usually, this is the inverse affine given by your nifti files or trk files (if the trk has been correctly generated)
- current_grid_to_world  # grid to world space / affine_matrix 
- stream_to_ref_grid   # # scanner to grid coordinate of your moving image. usually, this is the inverse affine given by your nifti files or trk files (if the trk has been correctly generated)
- ref_grid_to_world # grid to world space / affine_matrix.

the goal is to bring all the data in the same space so you need to know if you previous data were in the same space or different space like RAS+, LAS etc...

Also, I just discover that `deform_streamlines` has a bug. it will be fixed on the next release via PR  #2369

**Reply by skoudoro** [reply]
Not sure I am clear so let me know if you have any questions

**Reply by jcgriff1** [reply]
So, say that I:

- Rigidly aligned my anatomical to my DWI scan
- Nonlinearly warped the resulting aligned anatomical to the DWI 

I now have two transforms. The rigid transform maps from the anatomical to the DWI, and the non-linear warps the result to the DWI. 

Then, say that I:

- generate streamlines in DWI space, and I don't save them yet.
- want to use deform_streamlines to "unwarp" the streamlines to the aligned anatomical before transforming them back to anatomical space.

I'm not sure if this is the bug you're talking about, but I modified deform_streamlines to calculate new_streams_in_world like this: 

`    # Displace streamlines in world space
    new_streams_in_world = [
        np.array(np.array(d) + s) for d, s in zip(displacements, stream_in_world)
    ]
`

I've gotten reasonable results (i.e., the results visualize correctly on the anatomical) if I do this: 

`    # Apply the warp to the streamlines
    streamlines_anat = deform_streamlines(
        streamlines=streamlines,
        deform_field=anat_to_dwi_warp.get_forward_field(),  # non-linear transform; unclear why it's forward (but this is what works)
        stream_to_current_grid=np.eye(4),  # streamline to diffusion space
        current_grid_to_world=trk_file.affine,  # dwi space to world
        stream_to_ref_grid=np.linalg.inv(
            trk_file.affine
        ),  # this cancels out with the ref_grid_to_world
        ref_grid_to_world=trk_file.affine,  # this cancels out with stream_to_ref_grid
    )
    # Apply the inverse DWI affine
    streamlines_anat = transform_streamlines(
        streamlines_anat, np.linalg.inv(trk_file.affine)
    )
    # Apply the DWI image affine
    streamlines_anat = transform_streamlines(streamlines_anat, dwi_img.affine)
`
and then save the .trk file with the anat_to_dwi.affine as the affine.

I'm sure that this is wrong, but it works for some reason. 

What is the correct way to use this in this context?

**Reply by jcgriff1** [reply]
I think that I figured it out: 

`    trk_file_header_anat = helper.gen_trk_file_header(anat_img)
    # Apply the warp to the streamlines
    streamlines_anat = helper.deform_streamlines(
        streamlines=streamlines,
        deform_field=anat_to_dwi_warp.get_forward_field(),  # non-linear transform; unclear why it's forward (but this is what works)
        stream_to_current_grid=np.eye(4),  # streamline to diffusion space
        current_grid_to_world=dwi_img.affine,  # dwi space to world
        stream_to_ref_grid=anat_to_dwi.affine_inv,  # DWI to anatomical transform
        ref_grid_to_world=anat_img.affine,  # anatomical affine
    )`
    
 and then save the .trk file with the anat_img.affine.
    
 I still don't understand why it uses the get_forward_field() for the deformation, but this works and using the backward field doesn't.
